File: src/main/java/com/wxl/dyttcrawler/processor/AbstractDyttProcessor.java
package com.wxl.dyttcrawler.processor;

import com.wxl.dyttcrawler.core.DyttConstants;
import com.wxl.dyttcrawler.urlhandler.PriorityUrlCalculator;
import com.wxl.dyttcrawler.urlhandler.UrlFilter;
import org.springframework.util.StringUtils;
import us.codecraft.webmagic.Page;
import us.codecraft.webmagic.Request;
import us.codecraft.webmagic.utils.UrlUtils;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Create by wuxingle on 2020/5/16
 * 包含Link相关处理
 */
public abstract class AbstractDyttProcessor implements DyttProcessor {

    /**
     * url优先级计算
     */
    private PriorityUrlCalculator priorityCalculator;

    /**
     * url过滤器
     */
    private UrlFilter urlFilter;

    public AbstractDyttProcessor(PriorityUrlCalculator priorityCalculator,
                                 UrlFilter urlFilter) {
        this.priorityCalculator = priorityCalculator;
        this.urlFilter = urlFilter;
    }


    /**
     * 添加link
     */
    public void addLinks(Page page) {
        List<String> all = page.getHtml().links().all();
        all.stream().map(url -> canonicalizeUrl(page, url))
                .filter(url -> url != null && urlFilter.filter(url))
                .forEach(url -> addExtraLink(page, url));
    }

    /**
     * 添加link
     */
    public void addLink(Page page, String url) {
        URL url1 = canonicalizeUrl(page, url);
        if (url1 != null && urlFilter.filter(url1)) {
            addExtraLink(page, url1);
        }
    }

    /**
     * 存结果对象
     */
    public void putObject(Page page, Object obj) {
        if (obj != null) {
            page.putField(obj.getClass().getName(), obj);
        }
    }

    /**
     * url格式化
     *
     * @return 绝对地址
     */
    private URL canonicalizeUrl(Page page, String url) {
        if ("#".equals(url) || !StringUtils.hasText(url)) {
            return null;
        }
        url = UrlUtils.canonicalizeUrl(url, page.getUrl().toString());
        if (StringUtils.hasText(url)) {
            try {
                return new URL(url);
            } catch (MalformedURLException e) {
                return null;
            }
        }
        return null;
    }

    private void addExtraLink(Page page, URL url) {
        Request request = new Request();
        request.setUrl(url.toString());

        int priority = priorityCalculator.calculate(page, url);
        request.setPriority(priority);

        Integer depth = (Integer) page.getRequest().getExtra(DyttConstants.RequestAttr.DEPTH);

        Map<String, Object> extra = new HashMap<>();
        extra.put(DyttConstants.RequestAttr.DEPTH, depth == null ? 0 : depth + 1);

        request.setExtras(extra);

        page.addTargetRequest(request);
    }
}

File: src/main/kotlin/com/wxl/dyttcrawler/processor/AbstractDyttProcessor.kt
package com.wxl.dyttcrawler.processor

import com.wxl.dyttcrawler.core.RequestAttr
import com.wxl.dyttcrawler.urlhandler.PriorityUrlCalculator
import com.wxl.dyttcrawler.urlhandler.UrlFilter
import us.codecraft.webmagic.Page
import us.codecraft.webmagic.Request
import us.codecraft.webmagic.utils.UrlUtils
import java.net.MalformedURLException
import java.net.URL

/**
 * Create by wuxingle on 2021/10/12
 * 包含Link相关处理
 */
abstract class AbstractDyttProcessor(
    /**
     * url优先级计算
     */
    private val priorityUrlCalculator: PriorityUrlCalculator,

    /**
     * url过滤器
     */
    private val urlFilter: UrlFilter
) : DyttProcessor {


    /**
     * 添加link
     */
    fun addLinks(page: Page) {
        val all = page.html.links().all()
        all.mapNotNull { canonicalizeUrl(page, it) }
            .filter { urlFilter.match(it) }
            .forEach { addExtraLink(page, it) }
    }


    fun addLink(page: Page, url: String) {
        val newUrl = canonicalizeUrl(page, url)
        if (newUrl != null && urlFilter.match(newUrl)) {
            addExtraLink(page, newUrl)
        }
    }

    /**
     * 存结果对象
     */
    fun putObject(page: Page, obj: Any?) {
        if (obj != null) {
            page.putField(obj::class.java.name, obj)
        }
    }


    /**
     * url格式化
     *
     * @return 绝对地址
     */
    private fun canonicalizeUrl(page: Page, url: String?): URL? {
        if (url.isNullOrBlank() || "#" == url) {
            return null
        }
        val newUrl = UrlUtils.canonicalizeUrl(url, page.url.toString())
        if (newUrl.isNotBlank()) {
            try {
                return URL(newUrl)
            } catch (e: MalformedURLException) {
                //ignore
            }
        }
        return null
    }


    private fun addExtraLink(page: Page, url: URL) {
        val request = Request()
        request.url = url.toString()

        val priority = priorityUrlCalculator.calculate(page, url)
        request.priority = priority.toLong()

        val depth = page.request.getExtra(RequestAttr.DEPTH) as Int?
        request.extras = mutableMapOf<String, Any>(RequestAttr.DEPTH to (depth ?: 0))

        page.addTargetRequest(request)
    }


}

